### ISSUE TITLE
C++, link classes in :exception: info fields to the respective class documentation

### ISSUE THREAD

**Issue by MarcoKoch** [sphinx-doc/sphinx#2607]
I often have a setup like this:

```
.. cpp:class:: MyException : public std::exception

    Type of an exception

.. cpp:function:: void foo()

    :exception MyException: Something went wrong
```

With more complex interfaces, the exception class is often documented on an other page than the functions and classes that throw the exception. It would be nice if the mention of `MyException` in the documentation of `foo` would link to the documentation of `MyException`.

**Comment by jakobandersen** [comment]
Thanks for reporting, however when I try the example `MyException` becomes a link (same when splitting the declarations into separate pages). With which version of Sphinx do you observe the problem?

**Comment by MarcoKoch** [comment]
Ah, sorry. I tested my example code again and it actually gives me a link too. I'm using Sphinx 1.4.2.

I usually have my exception classes in a namespace, like `MyNamespace::MyException`. When documenting the exception in the `:exception:` info field, the double colons confuse Sphinx. I thus escape them like `:exception MyNamespace\:\:MyException:`.

Before writing the issue, I did a quick test without any namespaces to see whether the feature exists at all (which it obviously does) and only the escaped colons prevent the links. It seems like I had an unnoticed typo in my test code and thus did not get a link back then.

Is there any way to get the links with scoped exception classes too?

**Comment by jakobandersen** [comment]
Do you have an example where it fails? I tried the following example, and variations of it, but the exception field is properly converted into a link.

``` rst
.. cpp:class:: NS::MyException : public std::exception

    Type of an exception

.. cpp:function:: void foo()

    :exception NS\:\:MyException: Something went wrong
```

**Comment by MarcoKoch** [comment]
I tracked this down a bit. The problem I am experiencing seems to actually be that the `:cpp:namespace::` directive is ignored when the links are created:

``` rst
.. cpp:class:: NS::MyException : public std::exception

    Type of an exception.


.. cpp:namespace:: NS

.. cpp:function:: void my_function()

    :exception NS\:\:MyException:   This works.
    :exception MyException:         This doesn't.

    * Referencing :cpp:class:`NS::MyException` works.
    * Referencing :cpp:class:`MyException` also works.
```

**Comment by jakobandersen** [comment]
Thanks a lot for the example! I have fixed the part that made the following not work.

``` rst
.. cpp:class:: NS::MyException : public std::exception

.. cpp:function:: void NS::my_function()

    * Referencing :cpp:class:`NS::MyException` works.
    * Referencing :cpp:class:`MyException` didn't work before.
```

However, the bug you found extends to the other domains and probably requires some additional changes, so I have opened another issue for that part.

**Comment by jakobandersen** [comment]
Fixed with PR #3528.
